beacon-chain/sync/service.go

```diff
@@ -164,14 +167,14 @@ func NewService(ctx context.Context, opts ...Option) *Service {
 	r.rateLimiter = newRateLimiter(r.cfg.p2p)
 	r.initCaches()
 
-	go r.registerHandlers()
-	go r.verifierRoutine()
-
 	return r
 }
 
 // Start the regular sync service.
 func (s *Service) Start() {
+	go s.verifierRoutine()
+	go s.registerHandlers()
```

Comment:
Why did you move this from `NewService`?

Reply:
Thanks for noticing this, it's something I want input on. I think the intent of the `NewService` pattern is to centralize initialization of values, so that service dependencies can be wired up. Whereas the intent of `Start` is to actually kick off the Service goroutines and runloops, after services have been configured and initialized. `sync.Service` is unique in actually starting up goroutines to do meaningful work in `NewService`, which muddles the distinction. I assumed that starting these goroutines in NewService was a hack to deal with raciness and make sure sync was able to register well ahead of the start event being triggered, which is no longer an issue. @nisdas do you have any thoughts?

Reply:
I think in rare cases the initial sync service can send the `synced` event before the `sync` service starts up, if that happens the node will be stuck and unable to subscribe to gossip. To play on the safer side, the spawning off of the goroutines should be in the constructor for now. We can do something similar for the `synced` event in a follow up PR